core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -126,13 +148,23 @@ public int read(byte[] b, int off, int len) throws IOException {
 
     int bytesLeft = len;
     int currentOffset = off;
+    CountingRetry retry = new CountingRetry(MAX_WORKERS_TO_RETRY);
     while (bytesLeft > 0 && mPosition != mLength) {
       updateStream();
-      int bytesRead = mBlockInStream.read(b, currentOffset, bytesLeft);
-      if (bytesRead > 0) {
-        bytesLeft -= bytesRead;
-        currentOffset += bytesRead;
-        mPosition += bytesRead;
+      try {
+        int bytesRead = mBlockInStream.read(b, currentOffset, bytesLeft);
+        if (bytesRead > 0) {
+          bytesLeft -= bytesRead;
+          currentOffset += bytesRead;
+          mPosition += bytesRead;
+        }
+        retry.reset();
+      } catch (UnavailableException | ConnectException | DeadlineExceededException e) {
+        if (!retry.attempt()) {
```

Comment:
@bf8086 Please check if this still correct.

Reply:
This still works but the retry count semantic is changed. I will address it in a separate PR.